File: abc210/E/main.cpp
#include <algorithm>
#include <iostream>
using namespace std;

int main() {
  long N, M;
  cin >> N >> M;
  pair<long, long> edges[M];
  for (int i = 0; i < M; i++) {
    cin >> edges[i].second >> edges[i].first;
  }
  sort(edges, edges + M);

  long tot = 0;
  for (int i = 0; i < M; i++) {
    auto edge = edges[i];
    long c = edge.first, a = edge.second;
    long g = __gcd(a, N);
    tot += (N - g) * c;
    N = g;
  }

  if (N <= 1) {
    cout << tot << endl;
  } else {
    cout << -1 << endl;
  }
  return 0;
}

File: abc210/E/main.py
#!/usr/bin/env python3.8
# https://note.nkmk.me/python-union-find/
from math import gcd

N,M=map(int, input().split())
ns = set(range(N))
edges = []
for i in range(M):
  a,c=map(int, input().split())
  edges.append((c, a))

edges.sort()

tot = 0
xx = N
g = N
for edge in edges:
  c, a = edge
  g = gcd(a, g)
  xx -= g
  tot += xx * c

# print('xx', xx)
if xx <= 1:
  print(tot)
else:
  print(-1)